contrib/babelfishpg_tsql/src/pltsql_utils.c

```diff
@@ -2509,3 +2510,170 @@ get_owner_of_schema(const char *schema)
+		schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
+	}
+
+	/* Get physical schema name from logical schema name. */
+	physical_sch_name = get_physical_schema_name(db_name, schema_name);
+	/* Get namespace oid from physical schema name. */
+	obj_schema_oid = get_namespace_oid(physical_sch_name, false);
+
+	pfree(db_name);
+	pfree(schema_name);
+	pfree(physical_sch_name);
+
+	return obj_schema_oid;
+}
+
+/*
+ * get_proargtypes_oid
+ * Given a procedure name, namespace,
+ * user ID, and target argument name
+ * return the OID of the argument's type in the procedure.
+ *
+ * Returns InvalidOid if no matching procedure argument is found.
+ */
+static Oid
+get_proargtypes_oid(char *proname, Oid pronamespace, Oid user_id, char *targeted_arg_name)
+{
+	HeapTuple    tuple;
+	CatCList   *catlist;
+	Oid matched_type = InvalidOid;
+
+	/* Downcase and truncate identifier if needed. */
+	targeted_arg_name = downcase_truncate_identifier(targeted_arg_name, strlen(targeted_arg_name), true);
+
+	/* First search in pg_proc by name. */
+	catlist = SearchSysCacheList1(PROCNAMEARGSNSP, CStringGetDatum(proname));
+
+	for (int i = 0; i < catlist->n_members; i++)
+	{
+		Form_pg_proc procform;
+
+		tuple = &catlist->members[i]->tuple;
+		procform = (Form_pg_proc) GETSTRUCT(tuple);
+
+		/* Then consider only procs in specified namespace. */
+		if (procform->pronamespace == pronamespace &&
+			object_aclcheck(ProcedureRelationId, procform->oid, user_id, ACL_EXECUTE) == ACLCHECK_OK)
+		{
+			/* Get the list of proargames and corresponding proargtypes oids. */
+			char **proargnames = fetch_func_input_arg_names(tuple);
+			Oid *proargtypes = procform->proargtypes.values;
+
+			/* Find the typeoid corresponding to target TVP argument. */
+			for (int j = 0; j < procform->pronargs; j++)
+			{
+				if (strcmp(proargnames[j], targeted_arg_name) == 0)
+				{
+					matched_type = proargtypes[j];
+					break;
+				}
+			}
+		}
+	}
+	ReleaseSysCacheList(catlist);
+	if (matched_type == InvalidOid)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_FUNCTION),
+				errmsg("No procedure found with name \"%s\" that has an argument named \"%s\"",
+						proname, targeted_arg_name)));
+	}
+	return matched_type;
+}
+
+/*
+ * get_tvp_typename_typeschemaname
+ * Retrieves the type name and schema name of a Table-Valued Parameter (TVP)
+ * for a given stored procedure and argument name.
+ */
+void
+get_tvp_typename_typeschemaname(char *proc_name, char *target_arg_name, char **tvp_type_name, char **tvp_type_schema_name)
+{
+	bool			xactStarted = IsTransactionOrTransactionBlock();
+	Oid 			tvp_proargtype = InvalidOid;
+	Oid 			user_id = InvalidOid;
+	Oid 			obj_schema_oid = InvalidOid;
+	HeapTuple		tuple;
+	Oid 			typnamespace_oid = InvalidOid;
+	char 			*typnamespace;
+	char 			*curr_db;
+	MemoryContext 	oldContext;
+
+	if (!xactStarted)
+		StartTransactionCommand();
+	user_id = GetUserId();
+	curr_db = get_cur_db_name();
```

Comment:
In case of cross-db execution,
1. When we reach here what is the current_user -> TSQL user or TSQL login?
2. What is the current database context? Asking if we are switching the database before executing this procedure.

Reply:
We currentl do not support Cross-db execution for procedure having TVP as an argument. In `TdsRecvTypeTable` here , we use the current db as the default type database. 
https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_4_X_DEV/contrib/babelfishpg_tds/src/backend/tds/tdstypeio.c#L2323
And we do not allow the user to explicitly send the type database either.
https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_4_X_DEV/contrib/babelfishpg_tds/src/include/tds_request.h#L518